fix(qbft): prevent unsafe leader fallback when highest prepared data missing

diegomrsantos opened the pull request:
## Issue Addressed

Addresses one critical QBFT specification compliance issue from the broader work in PR #592.

Previously, when the leader had a highest prepared RoundChange but the full_data was missing, it would incorrectly fall back to proposing the start_data. This violates the QBFT specification which requires the leader to only propose from the highest prepared value when it exists.

## Proposed Changes

This change introduces `RcJustificationOutcome` enum to distinguish between three scenarios:
- `HighestPrepared(ValidData<D>)`: Valid prepared data available for proposal
- `PreparedExistsButDataMissing(Hash256)`: Prepared value exists but data is inaccessible  
- `NoPrepared`: No prepared values, safe to use start_data

When `PreparedExistsButDataMissing` occurs, the leader now correctly waits instead of unsafely proposing start_data, ensuring QBFT specification compliance.

### Key Changes:
1. **RcJustificationOutcome enum**: Replaces boolean return with explicit outcome states
2. **Leader logic update**: Handles missing prepared data by waiting instead of fallback
3. **Comprehensive test**: Verifies leader waits when highest prepared data is missing

## Additional Info

This is part of the broader QBFT specification compliance effort detailed in PR #592. For complete technical context and rationale, refer to that PR's description and comments.

This focused change addresses only the leader proposal fallback issue, with additional specification compliance fixes to follow in separate PRs.

diegomrsantos committed:
fix(qbft): prevent unsafe leader fallback when highest prepared data missing

Previously, when the leader had a highest prepared RoundChange but the full_data
was missing, it would incorrectly fall back to proposing the start_data. This
violates the QBFT specification which requires the leader to only propose from
the highest prepared value when it exists.

This change introduces RcJustificationOutcome enum to distinguish between:
- HighestPrepared: Valid prepared data available for proposal
- PreparedExistsButDataMissing: Prepared value exists but data is inaccessible
- NoPrepared: No prepared values, safe to use start_data

When PreparedExistsButDataMissing occurs, the leader now correctly waits instead
of unsafely proposing start_data, ensuring QBFT specification compliance.

Includes comprehensive test verifying the leader waits when highest prepared
data is missing.

Part of the broader QBFT spec compliance work detailed in PR #592.

diegomrsantos committed:
fix: address copilot feedback on test comments

- Convert doc comments to regular comments for test function
- Use consistent comment style throughout test code

diegomrsantos committed:
fmt

diegomrsantos commented on anchor/common/qbft/src/tests.rs:
```diff
@@ -356,3 +356,173 @@ fn test_round_change_validation_skips_round_one_prepared_values() {
+            root: prepared_hash,                // Claims this hash was prepared
+            data_round: 1,                      // Claims preparation happened in round 1
+            round_change_justification: vec![], // No RC justifications needed for this test
+            prepare_justification: vec![],      /* Should have prepare messages but we'll skip
+                                                 * validation */
```
fmt does that

Zacholme7 commented on anchor/common/qbft/src/lib.rs:
```diff
@@ -414,14 +423,29 @@ where
-
-            // Send the initial proposal and then the following prepare
-            self.send_proposal(valid_data.hash, valid_data.data.expect("Start data exists"));
+            match self.justify_round_change_quorum() {
```
could do something like 
```rust
let (hash, data) = match self.justify_round_change_quorum() {
// matching logic in here, return corresponding data or from function
}
self.send_proposal(hash, data);
```

diegomrsantos replied:
Done. Also removed the expects

diegomrsantos committed:
refactor(qbft): improve type system for message content handling

- Replace ValidData with Option<Arc<D>> pattern with dedicated types:
  - ValidData<D> for complete data (hash + Arc<D>)
  - HashOnly for hash-only references
  - MessageContent<D> enum combining both variants
- Implement requested let (hash, data) destructuring pattern in start_round()
- Eliminate expect() calls using safe pattern matching
- Add debug logging for ignored proposals without complete data
- Update validate_message return type to use MessageContent enum
- Maintain all existing functionality while improving type safety

Zacholme7 commented on anchor/common/qbft/src/lib.rs:
```diff
@@ -50,18 +50,44 @@ impl<D: QbftData<Hash = Hash256>> MessageData<D> {
+pub enum MessageContent<D: QbftData<Hash = Hash256>> {
+    Complete(ValidData<D>),
+    HashOnly(HashOnly),
+}
```
IMO I feel like it was more clean and simple before this change, this feels a bit over-engineering the problem

diegomrsantos replied:
We were forced to either call expect or handle the None case when there shouldn't be any None

Zacholme7 replied:
Got it

Zacholme7 commented on anchor/common/qbft/src/lib.rs:
```diff
@@ -50,18 +50,44 @@ impl<D: QbftData<Hash = Hash256>> MessageData<D> {
+#[derive(Debug, Clone)]
+pub enum MessageContent<D: QbftData<Hash = Hash256>> {
+    Complete(ValidData<D>),
+    HashOnly(HashOnly),
```
can we get rid of `HashOnly` and just use `Hash256` here?

diegomrsantos replied:
Done

diegomrsantos committed:
refactor(qbft): use generic D::Hash type in MessageContent and RcJustificationOutcome

- MessageContent::HashOnly now uses D::Hash instead of Hash256
- RcJustificationOutcome::PreparedExistsButDataMissing uses D::Hash
- Eliminates unnecessary HashOnly struct wrapper
- Makes code more generic and consistent with type constraints
- All pattern matching updated to use hash directly

diegomrsantos committed:
restore original parameter order

diegomrsantos commented on anchor/common/qbft/src/lib.rs:
```diff
@@ -50,14 +50,28 @@ impl<D: QbftData<Hash = Hash256>> MessageData<D> {
 
 impl<D: QbftData<Hash = Hash256>> ValidData<D> {
-    fn new(data: Option<Arc<D>>, hash: Hash256) -> Self {
+    fn new(hash: Hash256, data: Arc<D>) -> Self {
```
order restored

diegomrsantos commented on anchor/common/qbft/src/lib.rs:
```diff
@@ -50,14 +50,28 @@ impl<D: QbftData<Hash = Hash256>> MessageData<D> {
-    data: Option<Arc<D>>,
+    data: Arc<D>,
+}
+
```
data isn't Option anymore

diegomrsantos commented on anchor/common/qbft/src/lib.rs:
```diff
@@ -50,14 +50,28 @@ impl<D: QbftData<Hash = Hash256>> MessageData<D> {
+
+// Enum to represent message content - either complete data or just the hash
+#[derive(Debug, Clone)]
+pub enum MessageContent<D: QbftData<Hash = Hash256>> {
```
But I'm not happy with this name. @dknopik what does `Valid` mean in `ValidData`?

Zacholme7 replied:
Just that it passed `validate_message`

diegomrsantos replied:
But we don't do much validation in the `data` itself there, do we?

dknopik replied:
we do the value checks, i.e. whatever was passed here https://github.com/sigp/anchor/blob/65bce108d0abc881e7f9bfdc625a80aa40cb8e23/anchor/common/qbft/src/lib.rs#L143

Zacholme7 replied:
Not sure I am fully sold on the `MessageContent` name, but I unfortunately cannot think of something better

dknopik replied:
wdyt about `DataOrHash`?

Zacholme7 replied:
Im on board with that

diegomrsantos replied:
I'm not satisfied with the name either. But we can get this in as it is and create an issue to improve the naming in general, as imo it's a bit confusing:

```rust

pub trait QbftData: Debug + Clone + Encode + Decode {
    type Hash: Debug + Clone + Eq + Hash;

    fn hash(&self) -> Self::Hash;
}

pub trait QbftDataValidator<D: QbftData>: Send + Sync {
    fn validate(&self, value: &D, start_value: &D) -> bool;
}
```

wdyt?

Zacholme7 commented on anchor/common/qbft/src/lib.rs:
```diff
@@ -414,21 +430,32 @@ where
     pub fn receive(&mut self, wrapped_msg: WrappedQbftMessage) {
         // Perform base qbft releveant verification on the message
-        let Some((Some(valid_data), signer)) = self.validate_message(&wrapped_msg) else {
+        let Some((validated_msg, signer)) = self.validate_message(&wrapped_msg) else {
```
Maybe this? then we could remove the check below
```rust
let Some((MessageContent::Complete(valid_data), signer)) = self.validate_message(&wrapped_msg)
```

dknopik replied:
We cant do that as some message types do not have complete data. 

The check below is only on the `Proposal` variant

diegomrsantos committed:
fmt + fix

dknopik reviewed (approved):
LGTM, but feel free to apply the enum name change suggested above

diegomrsantos commented:
I've checked the Go code, and it does seem to behave like in this PR. Does anyone think it's worth it if someone else doubles check it?

dknopik commented:
done. It is kind of hard to assert, as they store things differently, but in the end, I believe the `data not found` case should not really happen anyway because of previous validation. Lets see. We'll stay vigilant and i'll try to catch any deviation in more fuzzing.
